core/src/docs/rfcs/3197_config.md

```diff
@@ -0,0 +1,234 @@
+}
+```
+
+Users can use `config()` to fetch the config of a service at runtime and construct a new operator based on needs.
+
+# Reference-level explanation
+
+Every services will have a `XxxConfig` struct.
+
+`XxxConfig` will implement the following things:
+
+- `Default` trait: All config fields will have a default value.
+- `Serialize` trait.
+- `Deserialize` trait.
+- `FromMap` trait: Convert from a `HashMap<String, String>`.
+- `Into<Config>` trait: All service config can be converted to `Config` enum.
+
+All config fields will be public and non-exhaustive, allowing users to build config this way:
+
+```rust
+let s3 = S3Config {
+  bucket: "test".to_string(),
+  endpoint: Some("http://localhost:9000".to_string()),
+  ..Default::default()
+}
+```
+
+The public API of existing builders will remain unchanged, although their internal implementations will be modified to utilize `XxxConfig`. Type that can't be represents as `String` like `Box<dyn AwsCredentialLoad>` and `HttpClient` will be kept in `Builder` as before.
+
+For example:
+
+```rust
+#[non_exhaustive]
+pub struct S3Config {
+  pub root: Option<String>,
+  pub bucket: String,
+  pub endpoint: Option<String>,
+  pub region: Option<String>,
+  ...
+}
+
+pub struct S3Builder {
+    config: S3Config,
+    
+    customed_credential_load: Option<Box<dyn AwsCredentialLoad>>,
+    http_client: Option<HttpClient>,
+}
+```
+
+# Drawbacks
+
+## API Surface
+
+This modification will significantly expand OpenDAL's public API surface, makes it harder to maintain and increases the risk of breaking changes. Also, this change will add much more work for bindings which need to implement `XxxConfig` for each service.
+
+## Secrets Leakage
+
+After our config supports `Serialize`, it's possible that users will serialize the config and log it. This will lead to secrets leakage. We should add a warning in the docs to prevent this. And we should also encourage uses of services like AWS IAM instead of static secrets.
+
+# Rationale and alternatives
+
+## Move `root` out of service config to operator level
+
+There is another way to solve the problem: [Move `root` out of service config to operator level](https://github.com/apache/incubator-opendal/issues/3151).
+
+We can move `root` out of the service config and put it in `Operator` level. This way, users can configure `root` for all services in one place. However, this is a large breaking changes and users will need to maintain the `root` logic everywhere.
+
+# Prior art
+
+None.
+
+# Unresolved questions
+
+None.
+
+# Future possibilities
+
+## Implement `FromStr` for `Config`
+
+We can implement `FromStr` for `Config` so that users can parse a config from a string.
+
+```rust
+let cfg = Config::from_str("s3://bucket/path/to/file?access_key_id=xxx&secret_access_key=xxx")?;
+```
+
+## Implement `Serialize` and `Deserialize` for `Config`
+
+We can implement `Serialize` and `Deserialize` for `Config` so that users can serialize and deserialize a config.
```

Comment:
For `Serialize`, we could skip some fields to keep them secret.

Reply:
I don't think we should. SKipping fields will break users appliations. THink about thise case:

The application reads user configurations and stores them in the application's metadata services using `Serialize`. If we omit certain fields, the application will break when attempting to read the configuration again.

Reply:
Maybe we'll have to explore how users would reuse this configuration information.
If we want to extract config from an operator and then create a new one based on this config, we can simply overwrite the fields that need to be modified.
For `Serialize`, reusing the config doesn't mean we need to know the full content of it. A secret field could always be secret.

Reply:
I don't think mark fields skipped works in this use case. But I do think Serialize needs more consideration. Let's move Serialize to future possibilities instead.

Reply:
Updated, please review agian.